lucene/core/src/java/org/apache/lucene/util/bkd/BKDWriter.java

```diff
@@ -730,6 +730,22 @@ private void rotateToTree(int nodeID, int offset, int count, byte[] index, List<
     return newLeafBlockFPs;
   }
 
+  private int getNumLeftLeaveNodes(int numLeaves) {
+    // return the max level for this number of leaves. If level is full it returns the next level
+    int maxLevel = 32 - Integer.numberOfLeadingZeros(numLeaves);
```

Comment:
is there an off-by-one error? I'd expect maxLevel to be `ceil(log2(numLeaves))`, which would be `32 - Integer.numberOfLeadingZeros(numLeaves-1)` according to https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/lang/Integer.html#numberOfLeadingZeros(int)

Reply:
I have rewritten the method (and fix some grammatical mistakes) to main more clear. I think the best way is to start calculating the full level which can be done using:

```
int lastFullLevel = 31 - Integer.numberOfLeadingZeros(numLeaves);
```

Then it is clear what we are doing.

Reply:
this makes sense to me